cachi2/core/package_managers/pip.py

```diff
@@ -1601,52 +1620,97 @@ def _validate_provided_hashes(requirements: list[PipRequirement], require_hashes
 
-    :return: Dict with package name, version and download path
-    :raises FetchError: if PyPI query failed
-    :raises PackageRejected: if sdists for the package is not found or yanked
-    """
-    timeout = get_config().requests_timeout
-    package = requirement.package
+def _process_package_distributions(
+    requirement: PipRequirement, pip_deps_dir: RootedPath, allow_binary: bool = False
+) -> tuple[DistributionPackageInfo, list[DistributionPackageInfo]]:
+    name = requirement.package
     version = requirement.version_specs[0][1]
+    normalized_version = canonicalize_version(version)
 
-    # See https://www.python.org/dev/peps/pep-0503/
-    package_url = f"{pypi_url.rstrip('/')}/simple/{canonicalize_name(package)}/"
+    client = pypi_simple.PyPISimple()
     try:
-        pypi_resp = pkg_requests_session.get(package_url, auth=pypi_auth, timeout=timeout)
-        pypi_resp.raise_for_status()
-    except requests.RequestException as e:
+        timeout = get_config().requests_timeout
+        project_page = client.get_project_page(name, timeout)
+        packages = project_page.packages
+    except (requests.RequestException, pypi_simple.NoSuchProjectError) as e:
         raise FetchError(f"PyPI query failed: {e}")
 
-    html = bs4.BeautifulSoup(pypi_resp.text, "html.parser")
-    # Find all anchors anywhere in the doc, the PEP does not specify where they should be
-    links = html.find_all("a")
+    def _to_dpi(dp: pypi_simple.DistributionPackage) -> DistributionPackageInfo:
+        dpi = DistributionPackageInfo(
+            name, version, pip_deps_dir.join_within_root(dp.filename).path, dp.url, dp.is_yanked
+        )
+        for algorithm, digest in dp.digests.items():
+            dpi.checksums.add(ChecksumInfo(algorithm, digest))
+        return dpi
+
+    wheels = []
+    if allow_binary:
+        filtered = [
+            p
+            for p in packages
+            if p.version
+            and canonicalize_version(p.version) == normalized_version
+            and p.package_type == "wheel"
+        ]
+
+        req_checksums = set(map(_to_checksum_info, requirement.hashes))
+        for wheel in filtered:
+            wheel_info = _to_dpi(wheel)
+            wheels.append(wheel_info)
+
+            both = wheel_info.checksums.intersection(req_checksums)
+            if both:
+                wheel_info.checksums = both
+                log.debug(
+                    "%s: using checksums reported by PyPI and specified by the user", wheel.filename
+                )
+
+            elif wheel_info.checksums:
+                log.debug("%s: using checksums reported by PyPI", wheel.filename)
+
+            elif req_checksums:
+                wheel_info.checksums = req_checksums
+                log.debug("%s: using checksums specified by the user", wheel.filename)
+
+            else:
+                wheels.pop()
+                log.info("Filtering out %s", wheel.filename)
```

Comment:
If I'm not mistaken, the code currently works like this

`if not wheel_info.checksums and not req_checksums` => we filter it out
`if wheel_info.checksums or req_checksums` => we don't filter it out

In other words: PyPI checksums are empty and user-specified checksums are empty <=> we filter it out

That's wrong for two reasons:
- if both are empty, we _do_ want to download the wheel
- if both are non-empty but have no intersection, we _don't_ want to download the wheel

Reply:
and if I try to put it in the `DistributionPackageInfo` ? Maybe it would be more extensible for future improvements